File: Algorithm/Kruskal's Algorithm/atcoder/A67 - MST (Minimum Spanning Tree)/A67.py
import sys
sys.setrecursionlimit(1 << 25)  # 再帰制限を緩めておく（Union-Findのfind用）

# 入力読み込み
input = sys.stdin.read
data = input().split()

N = int(data[0])
M = int(data[1])
edges = []

idx = 2
for _ in range(M):
    a = int(data[idx]) - 1
    b = int(data[idx + 1]) - 1
    c = int(data[idx + 2])
    edges.append((c, a, b))
    idx += 3

# 辺を重み順にソート
edges.sort()

# Union-Find構造体
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # 経路圧縮
        return self.parent[x]

    def union(self, x, y):
        rx = self.find(x)
        ry = self.find(y)
        if rx == ry:
            return False
        if self.rank[rx] < self.rank[ry]:
            self.parent[rx] = ry
        else:
            self.parent[ry] = rx
            if self.rank[rx] == self.rank[ry]:
                self.rank[rx] += 1
        return True

# クラスカル法でMST構築
uf = UnionFind(N)
total_cost = 0
edges_used = 0

for cost, u, v in edges:
    if uf.union(u, v):
        total_cost += cost
        edges_used += 1
        if edges_used == N - 1:
            break

print(total_cost)

File: Algorithm/Kruskal's Algorithm/atcoder/A67 - MST (Minimum Spanning Tree)/A67.js
const fs = require('fs');

const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const [N, M] = input[0].split(' ').map(Number);
const edges = [];

for (let i = 1; i <= M; i++) {
    const [a, b, c] = input[i].split(' ').map(Number);
    edges.push({ a: a - 1, b: b - 1, cost: c }); // 0-indexed
}

// コストの小さい順にソート
edges.sort((e1, e2) => e1.cost - e2.cost);

// Union-Find 構造体
class UnionFind {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = new Array(n).fill(0);
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]); // 経路圧縮
        }
        return this.parent[x];
    }

    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX === rootY) return false;

        // union by rank
        if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
        } else {
            this.parent[rootY] = rootX;
            if (this.rank[rootX] === this.rank[rootY]) {
                this.rank[rootX]++;
            }
        }
        return true;
    }
}

const uf = new UnionFind(N);
let totalCost = 0;
let edgesUsed = 0;

for (const edge of edges) {
    if (uf.union(edge.a, edge.b)) {
        totalCost += edge.cost;
        edgesUsed++;
        if (edgesUsed === N - 1) break; // MST 完成
    }
}

console.log(totalCost);
